slang-checks/src/main/java/com/sonarsource/slang/checks/NestedMatchCheck.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * SonarSource SLang
+ * Copyright (C) 2009-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package com.sonarsource.slang.checks;
+
+import com.sonarsource.slang.api.MatchCaseTree;
+import com.sonarsource.slang.api.MatchTree;
+import com.sonarsource.slang.api.Tree;
+import com.sonarsource.slang.checks.api.InitContext;
+import com.sonarsource.slang.checks.api.SlangCheck;
+import java.text.MessageFormat;
+import java.util.Optional;
+import org.sonar.check.Rule;
+
+@Rule(key = "S1821")
+public class NestedMatchCheck implements SlangCheck {
+  private static final String MESSAGE = "Refactor the code to eliminate this nested \"{0}\".";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(MatchTree.class, (ctx, matchTree) -> matchTree.cases().stream()
+      .map(NestedMatchCheck::getNestedMatchTrees)
+      .filter(Optional::isPresent)
+      .map(Optional::get)
+      .forEach(nestedMatch -> {
+          MatchTree match = (MatchTree) nestedMatch;
+          ctx.reportIssue(match.keyword().textRange(), MessageFormat.format(MESSAGE, match.keyword().text()));
+        }
+      ));
+  }
+
+  private static Optional<Tree> getNestedMatchTrees(MatchCaseTree matchCaseTree) {
+    return matchCaseTree
+      .descendants()
+      .filter(node -> node instanceof MatchTree)
+      .findFirst();
```

Comment:
I would return all nested `MatchTree` descendants instead of only the first one, then you could use `flatMap` and `distinct` on the returned stream to get single `MatchTree` for each violation.
Otherwise we might miss issue in the following case
```
when(x) {
  1 -> {
    when(y) { // OK -> issue raised correctly
      ...
    }
    when(z) { // KO -> no issue raised here
      ...
    }
  }
}
```

Reply:
Well spotted! I assumed each case being an expression, but that's not the case.